pkg/verification/verify.go

```diff
@@ -18,17 +18,175 @@ package verification
+	if opts.TsaCertificate == nil {
+		return nil
+	}
+
+	errMessage := ""
+
+	if !bytes.Equal(opts.TsaCertificate.RawIssuer, tsaCert.RawIssuer) {
+		errMessage += "TSR cert issuer does not match provided TSA cert issuer"
+	}
+
+	if opts.TsaCertificate.SerialNumber.Cmp(tsaCert.SerialNumber) != 0 {
+		if errMessage != "" {
+			errMessage += ", TSR cert issuer does not match provided TSA cert issuer"
+		} else {
+			errMessage = "TSR cert issuer does not match provided TSA cert issuer"
+		}
+	}
+
+	if errMessage != "" {
+		return errors.New(errMessage)
+	}
+	return nil
+}
+
+// Verify the leaf certificate's subject and/or subject alternative name matches a provided subject
+func verifyLeafCertSubject(cert *x509.Certificate, opts VerifyOpts) error {
+	if opts.Subject == "" {
+		return nil
+	}
+	
+	leafCertSubject := cert.Subject.String()
+	if leafCertSubject != opts.Subject {
+		return fmt.Errorf("Leaf cert subject %s does not match provided subject %s", leafCertSubject, opts.Subject)
+	}
+	return nil
+}
+
+// If embedded in the TSR, verify the TSR's leaf certificate matches a provided TSA certificate
+func verifyEmbeddedLeafCert(tsaCert *x509.Certificate, opts VerifyOpts) error {
+	if opts.TsaCertificate != nil && !opts.TsaCertificate.Equal(tsaCert) {
+		return fmt.Errorf("certificate embedded in the TSR does not match the provided TSA certificate")
+	}
+	return nil
+}
+
+func verifyLeafCert(ts timestamp.Timestamp, opts VerifyOpts) error {
+	if len(ts.Certificates) == 0 && opts.TsaCertificate == nil {
+		return fmt.Errorf("A leaf certificate must be present the in TSR or as a verify command argument")
+	}
+
+	errMsg := "failed to verify leaf cert"
+
+	var leafCert *x509.Certificate
+	if len(ts.Certificates) != 0 {
+		leafCert = ts.Certificates[0]
+		if opts.TsaCertificate != nil && !leafCert.Equal(opts.TsaCertificate) {
+			return fmt.Errorf("The leaf certificate included in the TSR does not match the one provied as a verify command argument")
+		}
+
+		err := verifyEmbeddedLeafCert(leafCert, opts)
+		if err != nil {
+			return fmt.Errorf("%s: %w", errMsg, err)
+		}
+	} else {
+		leafCert = opts.TsaCertificate
+	}
+
+	err := verifyESSCertID(leafCert, opts)
+	if err != nil {
+		return fmt.Errorf("%s: %w", errMsg, err)
+	}
+
+	err = verifyLeafCertSubject(leafCert, opts)
+	if err != nil {
+		return fmt.Errorf("%s: %w", errMsg, err)
+	}
+
+	return nil
+}
+
+func verifyExtendedKeyUsage(cert *x509.Certificate) error {
+	certEKULen := len(cert.ExtKeyUsage)
+	if certEKULen != 1 {
+		return fmt.Errorf("cert has %d extended key usages, expected only one", certEKULen)
+	}
+
+	if cert.ExtKeyUsage[0] != x509.ExtKeyUsageTimeStamping {
+		return fmt.Errorf("leaf cert EKU is not set to TimeStamping as required")
+	}
+	return nil
+}
+
+// Verify the TSA certificate and the intermediates (called "EKU chaining") all
+// have the extended key usage set to only time stamping usage
+func verifyLeafAndIntermediatesEKU(opts VerifyOpts) error {
```

Comment:
For only the leaf cert, you should also check the EKU is set to critical - https://github.com/sigstore/timestamp-authority/blob/main/pkg/x509/x509.go#L77-L83. Probably could check this in verifyLeafCert

(There's nothing in the RFC about criticality for intermediates)

Reply:
Added this in [ff25635](https://github.com/sigstore/timestamp-authority/pull/153/commits/ff2563516f66753fdeab898be4c6128c8f292194)